File: limbo/library-support/src/main/java/co/windly/limbo/activity/base/LimboActivity.java
package co.windly.limbo.activity.base;

import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import com.hannesdorfmann.mosby3.mvp.MvpActivity;

public abstract class LimboActivity<V extends LimboActivityView, P extends LimboActivityPresenter<V>>
  extends MvpActivity<V, P>
  implements LimboActivityView {

  //region Ui

  @LayoutRes
  protected abstract int getLayout();

  //endregion

  //region Lifecycle

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(getLayout());
    getPresenter().initializeManagers(this);
  }

  @Override
  protected void onDestroy() {
    getPresenter().clearDisposables();
    super.onDestroy();
  }

  //endregion
}

File: limbo/mvp/src/main/java/co/windly/limbo/mvp/activity/base/LimboActivity.kt
package co.windly.limbo.mvp.activity.base

import android.os.Bundle
import androidx.annotation.LayoutRes
import co.windly.limbo.mvp.LimboPresenter
import co.windly.limbo.mvp.LimboView
import com.hannesdorfmann.mosby3.mvp.MvpActivity
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

abstract class LimboActivity<V : LimboView, P : LimboPresenter<V>> : MvpActivity<V, P>(), LimboView {

  //region Reactive

  override val disposables: CompositeDisposable
    by lazy { CompositeDisposable() }

  override fun addDisposable(disposable: Disposable): Boolean =
    disposables.add(disposable)

  override fun clearDisposables() {
    disposables.clear()
  }

  //endregion

  //region Ui

  @get:LayoutRes
  protected abstract val layout: Int

  //endregion

  //region Lifecycle

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    // Define content view.
    setContentView(layout)
  }

  override fun onDestroy() {

    // Clear presenter-bound disposables.
    getPresenter().clearDisposables()

    // Clear view-bound disposables.
    clearDisposables()

    // Continue destroy'ing activity.
    super.onDestroy()
  }

  //endregion
}
